### PR TITLE
Extend rmdir of user storage to remove all contents of the bucket

### PR DESCRIPTION
**Pull request by IvanPombo**
This PR brings slight changes to the client to bring more awareness to the possibility of removing the full path and the errors that may occur.

In the backend, we change the endpoint to permit this possibility: see this PR inductiva/inductiva-web-api#582.

Minimal doubt: Should the removal of all content in the bucket have a prompt to confirm the user action? Or shall this be only applied in the CLI?

### CODE REVIEW

File: inductiva/tasks/run_simulation.py [hunk 1]
```diff
@@ -86,8 +86,9 @@ def run_simulation(
     logging.info(
         "Consider tracking the status of the task via CLI:"
         "\n\tinductiva tasks list --task-id %s", task_id)
-    logging.info("Or, tracking the logs of the task via CLI:"
-                 "\n\tinductiva logs %s", task_id)
+    logging.info(
+        "Or, tracking the logs of the task via CLI:"
+        "\n\tinductiva logs %s", task_id)
```
Comment (AugustoPeres) [inline inductiva/tasks/run_simulation.py:91]: Haha, I had to fix the same in my pull request :)

File: inductiva/resources/machines_base.py [hunk 2]
```diff
@@ -146,9 +146,10 @@ def start(self, **kwargs):
         try:
             self._api.start_vm_group(body=request_body)
         except inductiva.client.ApiException as e:
-            logs.log_and_exit(logging.getLogger(), logging.ERROR,
-                              "Starting machine group failed" 
-                              " with exception %s", e)
+            logs.log_and_exit(
```
Comment (AugustoPeres) [inline inductiva/resources/machines_base.py:149]: Why the change here?

File: inductiva/resources/machines_base.py [hunk 3]
```diff
@@ -146,9 +146,10 @@ def start(self, **kwargs):
...
+                logging.getLogger(), logging.ERROR,
```
Comment (AugustoPeres) [inline inductiva/resources/machines_base.py:150]: Same here. Was yapf failing on these files?

**Review by AugustoPeres — COMMENTED** [review]
I am not seeing the possibility of "removing contents of the bucket". It seems that we are just adding new logs, no? Am I missing something

File: inductiva/resources/machines_base.py (hunk 2 above)
Comment (IvanPombo) [reply inductiva/resources/machines_base.py:149]: Yapfing...

File: inductiva/resources/machines_base.py (hunk 3 above)
Comment (IvanPombo) [reply inductiva/resources/machines_base.py:150]: Yapfing... I am not sure what is going on to be honest...

**Comment by IvanPombo** [comment]
> I am not seeing the possibility of "removing contents of the bucket". It seems that we are just adding new logs, no? Am I missing something

This alters slightly the endpoint being used, but the structure of the method is exactly the same. Improve the docstring to refer that it can remove everything.

File: inductiva/storage/storage.py [hunk 4]
```diff
@@ -89,18 +91,51 @@ def _print_contents_table(contents):
...
     try:
         api = storage_api.StorageApi(inductiva.api.get_client())
-        api.delete_directory({"dir_name": path})
-        stripped_path = path.rstrip("/")
-        print(f"Directory deleted: {stripped_path}")
+        api.delete_path({"path": path})
+        logging.info("The path %s was successfully removed.", path)
     except inductiva.client.ApiException as api_exception:
+        if api_exception.status == 404:
+            raise api_exception.ValueError(f"Path {path} does not exist.")
```
Comment (sergiomsantos) [inline inductiva/storage/storage.py:137]:
```suggestion
            raise api_exception.ValueError(f"Unable to remove '{path}'. Path does not exit in user's remote storage.")
```

File: inductiva/storage/storage.py [hunk 5]
```diff
@@ -89,18 +91,51 @@ def _print_contents_table(contents):
...
+        elif api_exception.status == 500:
+            raise api_exception.RuntimeError(
+                f"Path {path} was failed to be removed.")
```
Comment (sergiomsantos) [inline inductiva/storage/storage.py:140]:
```suggestion
                f"Failed to remove remote path '{path}'.")
```

File: inductiva/storage/storage.py [hunk 6]
```diff
@@ -89,18 +91,51 @@ def _print_contents_table(contents):
...
+    else:
+        logging.info("Removing %s in the user storage.", path)
+
     try:
         api = storage_api.StorageApi(inductiva.api.get_client())
-        api.delete_directory({"dir_name": path})
-        stripped_path = path.rstrip("/")
-        print(f"Directory deleted: {stripped_path}")
+        api.delete_path({"path": path})
+        logging.info("The path %s was successfully removed.", path)
```
Comment (sergiomsantos) [inline inductiva/storage/storage.py:134]:
```suggestion
        logging.info("Successfully removed remote path '%s'.", path)
```

File: inductiva/storage/storage.py [hunk 7]
```diff
@@ -89,18 +91,51 @@ def _print_contents_table(contents):
...
+        confirm (bool): To perform the remove operation, the user must confirm
+             the operation by setting `confirm=True`. This is to avoid 
+             accidental removal of contents in the user's storage.
     """
+    if not confirm:
+        if path == "/":
+            path = "all contents"
+        raise RuntimeError(
+            "Please set `confirm=True` to confirm you want to"
+            f"delete {path} from the user's storage.")
```
Comment (sergiomsantos) [inline inductiva/storage/storage.py:124]:
```suggestion
            "Please set `confirm=True` to confirm you want to "
            f"delete {path} from the user's remote storage.")
```

File: inductiva/storage/storage.py [hunk 8]
```diff
@@ -89,18 +91,51 @@ def _print_contents_table(contents):
     )
 
 
-def rmdir(path: str):
-    """Deletes a directory inside the user's storage in Inductiva API.
-    Currently, directories are named by the task id of the task that
-    created them, so this function can be used to delete the directory
-    of a task.
+def rmdir(path: str, /, confirm: bool = False):
+    """Deletes the paths inside the user's storage in Inductiva API.
```
Comment (sergiomsantos) [inline inductiva/storage/storage.py:95]:
```suggestion
    """Delete a path from the user's remote storage in the Inductiva platform.
```

File: inductiva/storage/storage.py [hunk 9]
```diff
@@ -89,18 +91,51 @@ def _print_contents_table(contents):
...
+    `path="/"`. Note, this doesn't allow to remove specific files within 
+    any directory.
+
+    E.g.:
+        Works:
+        >>> inductiva.storage.rmdir("task_id", confirm=True)
+        >>> inductiva.storage.rmdir("/", confirm=True)
+
+        Fails:
+        >>> inductiva.storage.rmdir("task_id")
```
Comment (sergiomsantos) [inline inductiva/storage/storage.py]:
```suggestion
    This function removes the given `path` from the user's remote storage
    in the Inductiva platform. The given path can either be the id of a task to
    remove all content pertaining to that task, or `*` to remove all content.
    The function requires an explicit `confirm=True` keyword argument
    to confirm that the path is, indeed, to be removed.

    E.g.:
        #Works - remote path "task_id" exists and user is explicitly
        # sets confirm=True:
        >>> inductiva.storage.rmdir("task_id", confirm=True)
        
        #Works - The user explicitly sets confirm=True, confirming that
        # all contents in his remote storage space are to be removed:
        >>> inductiva.storage.rmdir("/", confirm=True)

        # Fails -  remote path "task_id" exists but the user fails to
        # confirm his intent. The call will fail win a runtime exception:
        >>> inductiva.storage.rmdir("task_id")
```

File: inductiva/storage/storage.py [hunk 10]
```diff
@@ -89,18 +91,51 @@ def _print_contents_table(contents):
...
+             accidental removal of contents in the user's storage.
     """
+    if not confirm:
+        if path == "/":
+            path = "all contents"
+        raise RuntimeError(
+            "Please set `confirm=True` to confirm you want to"
+            f"delete {path} from the user's storage.")
+
+    if path == "/":
```
Comment (sergiomsantos) [inline inductiva/storage/storage.py]:
```suggestion
    if path == "*":
```

File: inductiva/storage/storage.py [hunk 11]
```diff
@@ -89,18 +91,51 @@ def _print_contents_table(contents):
...
-        path (str): The path to the directory to delete.
+        path (str): Path relative to the root of the bucket to delete.
+            If path="/" then all the contents in the user's storage will be
+            removed.
+        confirm (bool): To perform the remove operation, the user must confirm
+             the operation by setting `confirm=True`. This is to avoid 
+             accidental removal of contents in the user's storage.
     """
+    if not confirm:
+        if path == "/":
```
Comment (sergiomsantos) [inline inductiva/storage/storage.py]:
```suggestion
        if path == "*":
```

File: inductiva/storage/storage.py [hunk 12]
```diff
@@ -89,18 +91,51 @@ def _print_contents_table(contents):
...
+        >>> inductiva.storage.rmdir("task_id")
+    
     Args:
-        path (str): The path to the directory to delete.
+        path (str): Path relative to the root of the bucket to delete.
+            If path="/" then all the contents in the user's storage will be
+            removed.
+        confirm (bool): To perform the remove operation, the user must confirm
+             the operation by setting `confirm=True`. This is to avoid 
+             accidental removal of contents in the user's storage.
```
Comment (sergiomsantos) [inline inductiva/storage/storage.py]:
```suggestion
        path (str): Path relative to the root of the bucket to delete.
            If path="*", all the contents in the user's remote storage space
            will be removed.
        confirm (bool): Confirmation flag to explicitly ensure the intent
            to remove the given path. The call will fail with a RuntimeException
            if the user does not set `confirm=True`.
```

**Review by sergiomsantos — CHANGES_REQUESTED** [review]
Some docstring clarification and changes in the way we refer to the entire content of the user's remote storage space.

File: inductiva/storage/storage.py (hunk 9 above)
Comment (IvanPombo) [reply inductiva/storage/storage.py]: Here, it is not necessarily the id of the task, since user's have flexibility to select the name of where a simulation is saved in storage.

**Review by sergiomsantos — APPROVED** [review]

**Comment by IvanPombo** [comment]
## QA Ready

To test the main functionalities of the storage remove methods you should use the dev environment by 
```bash
export INDUCTIVA_API_URL=https://api-dev.inductiva.ai
```

and your dev API key.

Notice, that this functionalities will remove elements from your storage, so please use carefully.

1. Remove a single task: First do `inductiva tasks list` to obtain a single `task_id` and then use python to remove the respective storage as follows:
```python
import inductiva

inductiva.storage.rmdir({task_id}, confirm=True)
```

Check that the task was removed from bucket with Python `inductiva.listdir(path={task_id})`.

2. Don't set `confirm=True` should fail the removal.

3. Remove everything in your user's bucket: 
```python
import inductiva

inductiva.storage.rmdir("*", confirm=True)
```

The last point may take awhile. In my case, I had 60 Gb of data and it took 1 min to delete everything, which is good.

Calling all my favorite QA testers: @AugustoPeres @Pbarbosa92 to delete your stuff.

**Comment by Pbarbosa92** [comment]
**QA**

Is this the expected result for removing a directory that does not exist ?
Captura de ecrã 2024-01-30, às 13 48 59

#########
The rmdir("*") works great.

**Comment by IvanPombo** [comment]
> **QA**
> 
> Is this the expected result for removing a directory that does not exist ? Captura de ecrã 2024-01-30, às 13 48 59
> 
> ######### The rmdir("*") works great.

Thanks @Pbarbosa92! I have fixed it now. Could you just briefly test again?

**Comment by Pbarbosa92** [comment]
> > **QA**
> > Is this the expected result for removing a directory that does not exist ? Captura de ecrã 2024-01-30, às 13 48 59
> > ######### The rmdir("*") works great.
> 
> Thanks @Pbarbosa92! I have fixed it now. Could you just briefly test again?

Can confirm, it works fine now ;)
